### DISCUSSION TITLE
Add autocomplete="off" to 2FA code input form

### DISCUSSION THREAD

**Discussion by Quix0r** [community/community#4346] [Other Feature Feedback, Questions, & Ideas] [answered]
Please consider adding a tiny `autocomplete="off"` to your 2FA code input form (when you have to enter the code) so it won't get stored in browser's form storage for better privacy.

Firefox 78.12.0esr (64-bit) under Debian Linux 11.

**Comment by TPS** [comment]
> for better privacy.

Rather, for irrelevance — saving the 2FA code is _pointless_, as it's 1-use _only_!  Also, it grows the browser storage uselessly.

**Reply by Quix0r** [reply]
Yes, and the code is always the same here strangely. But only for @github - For other sites I always get a new code. I use `oathtool`.

**Reply by Quix0r** [reply]
Confusing to me as well. Well, maybe I should make a screenshot of it when I enter the 2FA code.

**Comment by emmaviolet** [comment]
Just a note to say we're looking into this

**Comment by emmaviolet** [comment, marked as answer]
Just to follow up - we looked into this, but it looks like we'd lose some convenience for users who have SMS 2FA if we did this, and we'd rather not do that.

We're already using `autocomplete="one-time-code"`.

> In iOS and Safari on macOS we can [...] have the browser suggest two factor authentication codes that are sent to the device over SMS. Adding the `autocomplete` attribute with the value `"one-time-code"` will trigger this behaviour.

(https://www.twilio.com/blog/html-attributes-two-factor-authentication-autocomplete)

Therefore, we have two options, and it feels like "leave as-is" probably is the best:
1. leave as-is - get fancy SMS autocomplete on mac/iOS, get annoying autocomplete elsewhere
2. switch to `autocomplete="off"` - don't get fancy SMS autocomplete on mac/iOS, but also don't get annoying autocomplete elsewhere

**Reply by TPS** [reply]
3. Complain to Apple that they're doing it wrong & then proceed to hold breath 🤢

Seriously, can't OS/browser be detected & that rewritten dynamically for those who need it?

**Reply by emmaviolet** [reply]
OS can be detected, but then we'd still need to know whether users have received the code via SMS or not, which I guess we _could_ find, but would involve quite a round trip for a small change.

> Complain to Apple that they're doing it wrong & then proceed to hold breath 🤢

Yes, this was our first solution too 😇

**Reply by TPS** [reply]
> OS can be detected, but then we'd still need to know whether users have received the code via SMS or not, which I guess we _could_ find, but would involve quite a round trip for a small change.

Why so much? On 1st page/header request, detect OS/browser, then:  
- `autocomplete="off"` by default
- `autocomplete="one-time-code"` for any appropriate MacOS/Safari.

Why's SMS _detection_ relevant for this algo?

**Reply by emmaviolet** [reply]
> Why's SMS detection relevant for this algo?

There are multiple 2FA methods, so if you don't use SMS 2FA, presume you wouldn't want `autocomplete="one-time-code"` either.

**Reply by TPS** [reply]
> There are multiple 2FA methods, so if you don't use SMS 2FA, presume you wouldn't want `autocomplete="one-time-code"` either.

This seems like a case of perfect getting in the way of good enough for now.  If MacOS folks are the issue, let them have it as it is, but solve for everyone else sooner, then iterate when a better solution presents, no?

**Reply by emmaviolet** [reply]
> This seems like a case of perfect getting in the way of good enough for now. If MacOS folks are the issue, let them have it as it is, but solve for everyone else sooner, then iterate when a better solution presents, no?

Yeah, that's fair. To be honest your proposal seemed a little like a case of perfect getting in the way of good enough for now on first glance to me, too. I remain a little concerned that 1) our methods of detecting OS aren't perfectly reliable, and I'd rather not sacrifice convenience for the sake of irrelevance; and 2) that it'll involve js which feels kind of clunky for something like this.

But we're having a look to see whether that's justified ^^

Also, if you want to add your voice to the related firefox bug report... https://bugzilla.mozilla.org/show_bug.cgi?id=1547294

**Comment by Quix0r** [comment]
Okay, so does this mean `one-time-code` is some kind of non-standard value? Because Firefox follows W3C (as I know). I'm using 78.12.0esr (64-bit) here.

**Reply by TPS** [reply]
Per the link above, this behavior looks to be _(currently?)_ an Apple-only addition to the standard.

**Reply by Quix0r** [reply]
Well, okay. Before we continue bash Apple, let's close this discussing. Maybe your proposed change may work, default: `off`, MacOS/iOS detected: `one-time-code`.

**Comment by emmaviolet** [comment]
Apologies, was off last week, so this slipped a bit. Upshot, though, is that we looked into a js solution and it wasn't as clunky as we feared. Hope to ship soon.

**Reply by emmaviolet** [reply]
This shipped 4 days ago. Thanks for the pointer @Quix0r and to all of you for your feedback

**Reply by TPS** [reply]
@Quix0r Close FTW?

**Reply by Quix0r** [reply]
Thank you, I have enabled it already some time ago.

**Reply by TPS** [reply]
@Quix0r No, I mean, you've the ability to close this discussion, now that its RFE is implemented.

**Comment by TPS** [comment]
Obligatory xkcd:
![never_told_anyone_2x.c-o-d.png](https://user-images.githubusercontent.com/185902/142728948-994fe3b6-1707-4df1-bbb6-cc5287ae6d3d.png)
